File: lib/db/queries/tags.ts
// Tags: list, edit, delete. Create / assign in mutations.ts.

import type { Exec } from '@/lib/db/repo';

export interface Tag {
  id: string;
  ledgerId: string;
  name: string;
  color: string | null;
}

/** List tags; pass a ledgerId to scope, or omit for all ledgers. */
export async function listTags(exec: Exec, ledgerId?: string): Promise<Tag[]> {
  const rows = await exec(
    ledgerId ? 'SELECT * FROM tags WHERE ledger_id = ? ORDER BY name' : 'SELECT * FROM tags ORDER BY ledger_id, name',
    ledgerId ? [ledgerId] : [],
  );
  return rows.map((r) => ({
    id: String(r.id),
    ledgerId: String(r.ledger_id),
    name: String(r.name),
    color: r.color == null ? null : String(r.color),
  }));
}

export interface TagPatch {
  name?: string;
  color?: string | null;
}

/** Update a tag's editable fields. */
export async function updateTag(exec: Exec, id: string, patch: TagPatch): Promise<void> {
  const sets: string[] = [];
  const bind: (string | number | null)[] = [];
  if (patch.name !== undefined) { sets.push('name = ?'); bind.push(patch.name); }
  if (patch.color !== undefined) { sets.push('color = ?'); bind.push(patch.color ?? null); }
  if (!sets.length) return;
  sets.push("updated_at = datetime('now')");
  bind.push(id);
  await exec(`UPDATE tags SET ${sets.join(', ')} WHERE id = ?`, bind);
}

/** Hard delete a tag; transaction_tags rows cascade away via the FK. */
export async function deleteTag(exec: Exec, id: string): Promise<void> {
  await exec('DELETE FROM tags WHERE id = ?', [id]);
}

File: ios/FinchCore/Sources/FinchCore/Store/Domain/Tags.swift
import Foundation
import GRDB

/// Tags domain — port of lib/db/domain/tags/mutations.ts (createTag/updateTag/deleteTag).
public enum Tags {
    public static let handlers: [ActionName: Apply.Handler] = [
        .createTag: createTag,
        .updateTag: updateTag,
        .deleteTag: deleteTag,
    ]

    static func createTag(_ db: Database, _ args: Args) throws {
        struct A: Decodable { let id: String?; let ledgerId: String?; let name: String; let color: String? }
        let a = try args.to(A.self)
        let name = a.name.trimmingCharacters(in: .whitespacesAndNewlines)
        if name.isEmpty { throw I18nError("error.required.tagName", [:], "Tag name is required") }
        try db.execute(sql: "INSERT INTO tags (id,ledger_id,name,color,created_at,updated_at) VALUES (?,?,?,?,datetime('now'),datetime('now'))",
                       arguments: [a.id ?? Entries.newId("tag"), a.ledgerId ?? "personal", name, a.color])
    }

    static func updateTag(_ db: Database, _ args: Args) throws {
        guard case .string(let id)? = args.values["id"] else { throw I18nError("error.invalidArgs", [:], "updateTag requires an id") }
        let patch: [String: JSONValue] = { if case .object(let p)? = args.values["patch"] { return p }; return [:] }()
        var sets: [String] = []
        var bind: [DatabaseValueConvertible?] = []
        if let v = patch["name"] {
            guard case .string(let s) = v, !s.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                throw I18nError("error.required.tagName", [:], "Tag name is required")
            }
            sets.append("name = ?"); bind.append(s)
        }
        if patch.keys.contains("color") {
            if case .string(let s)? = patch["color"] { sets.append("color = ?"); bind.append(s) }
            else { sets.append("color = ?"); bind.append(nil as String?) }
        }
        if sets.isEmpty { return }
        sets.append("updated_at = datetime('now')")
        bind.append(id)
        try db.execute(sql: "UPDATE tags SET \(sets.joined(separator: ", ")) WHERE id = ?", arguments: StatementArguments(bind))
    }

    static func deleteTag(_ db: Database, _ args: Args) throws {
        struct A: Decodable { let id: String }
        try db.execute(sql: "DELETE FROM tags WHERE id = ?", arguments: [try args.to(A.self).id])
    }
}
